src/local-mode.ts

```diff
@@ -56,51 +102,150 @@ export class LocalModeManager {
-			const containers = await this.getLocalModeContainers(images);
-
-			await Bluebird.map(containers, containerId => {
-				log.debug('Removing local mode container: ', containerId);
-				return this.docker.getContainer(containerId).remove({ force: true });
-			});
-			await Bluebird.map(images, imageId => {
-				log.debug('Removing local mode image: ', imageId);
-				return this.docker.getImage(imageId).remove({ force: true });
-			});
-
-			// Remove any local mode state added to the database
-			await this.db
-				.models('app')
-				.del()
-				.where({ source: 'local' });
-		} catch (e) {
-			log.error('There was an error clearing local mode artifacts: ', e);
-		}
+	// Query the engine to get currently running containers and installed images.
+	public async collectEngineSnapshot(): Promise<EngineSnapshotRecord> {
+		const containersPromise = this.docker
+			.listContainers()
+			.then(resp => _.map(resp, 'Id'));
+		const imagesPromise = this.docker
+			.listImages()
+			.then(resp => _.map(resp, 'Id'));
+		const volumesPromise = this.docker
+			.listVolumes()
+			.then(resp => _.map(resp.Volumes, 'Name'));
+		const networksPromise = this.docker
+			.listNetworks()
+			.then(resp => _.map(resp, 'Id'));
+
+		const data = await Bluebird.all([
+			containersPromise,
+			imagesPromise,
+			volumesPromise,
+			networksPromise,
+		]);
+		return new EngineSnapshotRecord(
+			new EngineSnapshot(data[0], data[1], data[2], data[3]),
+			new Date(),
+		);
 	}
 
-	private async getLocalModeImages(): Promise<string[]> {
-		// Return all local mode images present on the local docker daemon
-		return _.map(
-			await this.docker.listImages({
-				filters: { label: ['io.resin.local.image=1'] },
-			}),
-			'Id',
+	// Store engine snapshot data in the local database.
+	public async storeEngineSnapshot(record: EngineSnapshotRecord) {
+		const timestamp = record.timestamp.toISOString();
+		log.debug(
+			`Storing engine snapshot in the database. Timestamp: ${timestamp}`,
 		);
+		await this.db.models('engineSnapshot').delete();
+		return this.db.models('engineSnapshot').insert({
+			snapshot: JSON.stringify(record.snapshot),
+			timestamp,
+		});
 	}
 
-	private async getLocalModeContainers(
-		localModeImageIds: string[],
-	): Promise<string[]> {
-		return _(await this.docker.listContainers())
-			.filter(({ Image }) => _.includes(localModeImageIds, Image))
-			.map('Id')
-			.value();
+	// Read the latest stored snapshot from the database.
+	public async retrieveLatestSnapshot(): Promise<EngineSnapshotRecord | null> {
+		const r = await this.db
+			.models('engineSnapshot')
+			.select()
+			.orderBy('rowid', 'DESC')
+			.first();
+
+		if (!r) {
+			return null;
+		}
+		return new EngineSnapshotRecord(
+			EngineSnapshot.fromJSON(r.snapshot),
+			new Date(Date.parse(r.timestamp)),
+		);
+	}
+
+	private async removeLocalModeArtifacts(objects: EngineSnapshot) {
+		log.debug(`Going to delete the following objects: ${objects}`);
```

Comment:
I do worry about how this will look with something like bob for example, that has a bunch of services and volumes.

Reply:
Are you worried about logs pollution?

Can replace this with counters in logs. Specific object deletion behaviour is covered with tests, so counters here should be fine.

Reply:
By counters do you mean `removing n images, n networks` etc?

Reply:
Yep, exactly.

Reply:
SGTM

Reply:
Done.